cmd/validate_asserter_config.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2020 Coinbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/coinbase/rosetta-sdk-go/asserter"
+	"github.com/coinbase/rosetta-sdk-go/fetcher"
+	"github.com/coinbase/rosetta-sdk-go/types"
+	"github.com/coinbase/rosetta-sdk-go/utils"
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// Common helper across Construction and Data
+// Issues an RPC to fetch /network/options, and extracts the `Allow`
+// Reads the JSON file at `asserterConfigurationFile` and loads into a Go object
+// Validates the `Allow`s across both objects match
+func validateNetworkOptionsMatchesAsserterConfiguration(
+	ctx context.Context, f *fetcher.Fetcher, network *types.NetworkIdentifier,
+	asserterConfigurationFile string,
+) error {
+	var asserterConfiguration asserter.Configuration
+	if err := utils.LoadAndParse(asserterConfigurationFile, &asserterConfiguration); err != nil {
+		return fmt.Errorf("%w: failure loading / parsing asserter-configuration-file", err)
+	}
+
+	resp, fetchErr := f.NetworkOptions(ctx, network, nil)
+	if fetchErr != nil {
+		return fmt.Errorf("%w: failure getting /network/options", fetchErr.Err)
+	}
+
+	return validateNetworkAndAsserterAllowMatch(resp.Allow, &asserterConfiguration)
+}
+
+func validateNetworkAndAsserterAllowMatch(
+	networkAllow *types.Allow, asserterConfiguration *asserter.Configuration,
+) error {
+	if networkAllow == nil {
+		return errors.New("/network/options object's Allow is nil")
+	}
+	if asserterConfiguration == nil {
+		return errors.New("asserter-configuration-file object is nil")
+	}
+
+	if err := verifyTimestampStartIndex(
+		networkAllow.TimestampStartIndex, asserterConfiguration.AllowedTimestampStartIndex,
+	); err != nil {
+		return err
+	}
+
+	if err := verifyOperationTypes(
+		networkAllow.OperationTypes, asserterConfiguration.AllowedOperationTypes,
+	); err != nil {
+		return err
+	}
+
+	if err := verifyOperationStatuses(
+		networkAllow.OperationStatuses, asserterConfiguration.AllowedOperationStatuses,
+	); err != nil {
+		return err
+	}
+
+	return verifyErrors(networkAllow.Errors, asserterConfiguration.AllowedErrors)
+}
+
+func verifyTimestampStartIndex(networkTsi *int64, assertTsi int64) error {
+	if networkTsi == nil { // This field is optional and defaults to all allowed
+		return nil
+	}
```

Comment:
I think we should expect the timestamp index to stay the same? i.e. if it is nil/0 in the static file, then it should be nil/0 in network/options? I think we can treat nils and 0s the same here, since both 0 and nil should mean timestamp for all indexes are valid.

Reply:
That's what I thought too but unfortunately it doesn't seem to default to zero
If I change this code to
```
	var networkTsiVal int64 // default to zero
	if networkTsi != nil { // This field is optional and defaults to all allowed
		networkTsiVal = *networkTsi
	}
	if networkTsiVal != assertTsi {
```
it hits an error
```
$ go run main.go --configuration-file ../rosetta-cli-configs/cardano.json check:data --asserter-configuration-file /Users/raghav.sharma@coinbase.com/go/src/rosetta-cli/asserterConfiguration.json
loaded configuration file: ../rosetta-cli-configs/cardano.json
Using temporary directory /var/folders/jt/dznrb0v172s_2m6718qtb8x40000gn/T/488414326

Error: /network/options / asserter-configuration-file timestamp start index mismatch. 0 1
```
because the network options is nil, but the asserter configuration file has 1

So defaulting to zero seems incorrect
Maybe we could default to 1, but I'm not sure if that will work for all blockchains. For example, maybe other blockchains use zero-indexing so nil should map to zero for those

Let me know what you think. Maybe just adding a comment explaining this should be sufficient